File: apps/mobile/ios/Storage/SchemeCacheService.swift
import CryptoKit
import Foundation

struct SchemeCacheResult {
    let updated: Bool
    let hasScheme: Bool
    let version: String
    let hash: String
}

final class SchemeCacheService {
    private let gateway: MobileGatewayClient
    private let logs: MobileLogStore
    private let fileManager: FileManager
    private let root: URL

    init(gateway: MobileGatewayClient, logs: MobileLogStore, fileManager: FileManager = .default) throws {
        self.gateway = gateway
        self.logs = logs
        self.fileManager = fileManager
        let support = try fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true)
        root = support.appendingPathComponent("OneDesk", isDirectory: true)
        try fileManager.createDirectory(at: root, withIntermediateDirectories: true)
    }

    func get(desktopId: String) -> JSONObject? {
        let url = cacheURL(desktopId: desktopId)
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONSupport.object(from: data)
    }

    func downloadAndCache(
        desktop: KnownDesktop,
        credential: String,
        descriptor: JSONObject
    ) throws -> SchemeCacheResult {
        let version = descriptor.string("version").isEmpty ? "0" : descriptor.string("version")
        let hash = descriptor.string("hash")
        let hasScheme = descriptor.bool("hasScheme")
        if !hash.isEmpty, get(desktopId: desktop.desktopId)?.string("hash") == hash {
            return SchemeCacheResult(updated: false, hasScheme: hasScheme, version: version, hash: hash)
        }

        var payload: JSONObject = hasScheme
            ? try downloadSchemePayload(desktop: desktop, credential: credential, descriptor: descriptor)
            : [
                "activeSchemeId": NSNull(),
                "scheme": NSNull(),
                "pages": [],
                "components": [],
                "actions": [],
                "permissionGrants": [],
            ]
        let assetRoot = assetDirectory(desktopId: desktop.desktopId, hash: hash)
        do {
            try fileManager.createDirectory(at: assetRoot, withIntermediateDirectories: true)
            try materializeAssets(desktop: desktop, credential: credential, root: assetRoot, payload: &payload)
        } catch {
            try? fileManager.removeItem(at: assetRoot)
            throw error
        }

        let cache: JSONObject = [
            "desktopId": desktop.desktopId,
            "version": version,
            "hash": hash,
            "updatedAt": ISO8601DateFormatter().string(from: Date()),
            "activeSchemeId": payload["activeSchemeId"] ?? NSNull(),
            "scheme": payload["scheme"] ?? NSNull(),
            "pages": payload["pages"] ?? [],
            "components": payload["components"] ?? [],
            "actions": payload["actions"] ?? [],
            "permissionGrants": payload["permissionGrants"] ?? [],
        ]
        // Data.write(.atomic) 先写同目录临时文件再替换，进程中断不会破坏上一个可用缓存。
        try JSONSupport.data(cache).write(to: cacheURL(desktopId: desktop.desktopId), options: .atomic)
        removeStaleAssetDirectories(desktopId: desktop.desktopId, currentHash: hash)
        return SchemeCacheResult(updated: true, hasScheme: hasScheme, version: version, hash: hash)
    }

    private func downloadSchemePayload(
        desktop: KnownDesktop,
        credential: String,
        descriptor: JSONObject
    ) throws -> JSONObject {
        let totalBytes = descriptor.int64("totalBytes")
        let expectedHash = descriptor.string("hash")
        guard totalBytes > 0, totalBytes <= 32 * 1024 * 1024, !expectedHash.isEmpty else {
            throw OneDeskMobileError.integrity("方案快照大小或哈希无效")
        }
        var output = Data(capacity: Int(totalBytes))
        var offset: Int64 = 0
        while offset < totalBytes {
            var request = gateway.authorizedRequest(type: "scheme-chunk", credential: credential)
            request["hash"] = expectedHash
            request["offset"] = offset
            request["length"] = 24 * 1024
            let response = try gateway.request(
                host: desktop.host,
                port: desktop.port,
                payload: request,
                expectedFingerprint: desktop.gatewayFingerprint)
            guard response.bool("ok"), let chunk = response.object("payload"),
                  chunk.int64("offset", default: -1) == offset,
                  let bytes = Data(base64Encoded: chunk.string("data")), !bytes.isEmpty else {
                throw OneDeskMobileError.integrity(response.string("message").isEmpty
                    ? "方案分块顺序无效" : response.string("message"))
            }
            output.append(bytes)
            offset += Int64(bytes.count)
        }
        guard sha256(output) == expectedHash.lowercased() else {
            throw OneDeskMobileError.integrity("方案完整性校验失败")
        }
        return try JSONSupport.object(from: output)
    }

    private func materializeAssets(
        desktop: KnownDesktop,
        credential: String,
        root: URL,
        payload: inout JSONObject
    ) throws {
        var pages = payload["pages"] as? [JSONObject] ?? []
        for index in pages.indices {
            try replaceMediaSource(
                desktop: desktop,
                credential: credential,
                root: root,
                ownerKind: "page",
                ownerId: pages[index].string("id"),
                target: &pages[index],
                key: "backgroundMediaSource")
        }
        payload["pages"] = pages

        var components = payload["components"] as? [JSONObject] ?? []
        for index in components.indices {
            let ownerId = components[index].object("definition")?.string("id") ?? ""
            guard var visual = components[index].object("visualConfig") else { continue }
            if var background = visual.object("background") {
                try replaceMediaSource(
                    desktop: desktop,
                    credential: credential,
                    root: root,
                    ownerKind: "component",
                    ownerId: ownerId,
                    target: &background,
                    key: "mediaSource")
                visual["background"] = background
            }
            if var image = visual.object("image") {
                try replaceMediaSource(
                    desktop: desktop,
                    credential: credential,
                    root: root,
                    ownerKind: "component",
                    ownerId: ownerId,
                    target: &image,
                    key: "source")
                visual["image"] = image
            }
            components[index]["visualConfig"] = visual
        }
        payload["components"] = components
    }

    private func replaceMediaSource(
        desktop: KnownDesktop,
        credential: String,
        root: URL,
        ownerKind: String,
        ownerId: String,
        target: inout JSONObject,
        key: String
    ) throws {
        guard let source = target[key] as? String,
              !source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !ownerId.isEmpty else { return }
        let fileName = URL(string: source)?.lastPathComponent.isEmpty == false
            ? URL(string: source)!.lastPathComponent
            : URL(fileURLWithPath: source).lastPathComponent
        guard !fileName.isEmpty else { return }
        do {
            let local = try downloadAsset(
                desktop: desktop,
                credential: credential,
                root: root,
                ownerKind: ownerKind,
                ownerId: ownerId,
                fileName: fileName)
            target[key] = local.absoluteString
        } catch {
            logs.append("Error", "SchemeAsset", "\(ownerKind)/\(ownerId)/\(fileName)：\(error.localizedDescription)")
            throw error
        }
    }

    private func downloadAsset(
        desktop: KnownDesktop,
        credential: String,
        root: URL,
        ownerKind: String,
        ownerId: String,
        fileName: String
    ) throws -> URL {
        let safeName = sanitize("\(ownerKind)-\(ownerId)-\(URL(fileURLWithPath: fileName).lastPathComponent)")
        let destination = root.appendingPathComponent(safeName)
        if let size = try? destination.resourceValues(forKeys: [.fileSizeKey]).fileSize, size > 0 {
            return destination
        }
        let temporary = root.appendingPathComponent("\(safeName).tmp")
        fileManager.createFile(atPath: temporary.path, contents: nil)
        let output = try FileHandle(forWritingTo: temporary)
        defer { try? output.close() }
        var offset: Int64 = 0
        var total = Int64.max
        while offset < total {
            var request = gateway.authorizedRequest(type: "asset", credential: credential)
            request["ownerKind"] = ownerKind
            request["ownerId"] = ownerId
            request["fileName"] = fileName
            request["offset"] = offset
            request["length"] = 24 * 1024
            let response = try gateway.request(
                host: desktop.host,
                port: desktop.port,
                payload: request,
                timeoutMilliseconds: 12_000,
                expectedFingerprint: desktop.gatewayFingerprint)
            guard response.bool("ok"), let chunk = response.object("payload"),
                  chunk.int64("offset", default: -1) == offset,
                  let bytes = Data(base64Encoded: chunk.string("data")) else {
                throw OneDeskMobileError.integrity(response.string("message").isEmpty
                    ? "资源分块顺序无效" : response.string("message"))
            }
            total = chunk.int64("totalBytes")
            if bytes.isEmpty && offset < total {
                throw OneDeskMobileError.integrity("资源分块为空")
            }
            try output.write(contentsOf: bytes)
            offset += Int64(bytes.count)
            if chunk.bool("complete") { break }
        }
        if fileManager.fileExists(atPath: destination.path) { try fileManager.removeItem(at: destination) }
        try fileManager.moveItem(at: temporary, to: destination)
        return destination
    }

    private func cacheURL(desktopId: String) -> URL {
        root.appendingPathComponent("scheme-\(sanitize(desktopId)).json")
    }

    private func assetDirectory(desktopId: String, hash: String) -> URL {
        root.appendingPathComponent("scheme-assets", isDirectory: true)
            .appendingPathComponent(sanitize(desktopId), isDirectory: true)
            .appendingPathComponent(sanitize(hash.isEmpty ? "empty" : hash), isDirectory: true)
    }

    private func removeStaleAssetDirectories(desktopId: String, currentHash: String) {
        let desktopRoot = root.appendingPathComponent("scheme-assets", isDirectory: true)
            .appendingPathComponent(sanitize(desktopId), isDirectory: true)
        guard let directories = try? fileManager.contentsOfDirectory(
            at: desktopRoot,
            includingPropertiesForKeys: [.isDirectoryKey]) else { return }
        let keep = sanitize(currentHash.isEmpty ? "empty" : currentHash)
        for directory in directories where directory.lastPathComponent != keep {
            try? fileManager.removeItem(at: directory)
        }
    }

    private func sanitize(_ value: String) -> String {
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-")
        return value.unicodeScalars.map { allowed.contains($0) ? Character(String($0)) : "_" }.reduce("") { $0 + String($1) }
    }

    private func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

File: apps/mobile/android/app/src/main/java/cc/onedesk/mobile/SchemeCacheService.kt
package cc.onedesk.mobile

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.time.Instant

data class SchemeCacheResult(val updated: Boolean, val hasScheme: Boolean, val version: String, val hash: String)

class SchemeCacheService(
    private val context: Context,
    private val prefs: SharedPreferences,
    private val gateway: MobileGatewayClient,
    private val logs: MobileLogStore,
) {
    fun get(desktopId: String): JSONObject? {
        val raw = prefs.getString("scheme:$desktopId", null) ?: return null
        return try { JSONObject(raw) } catch (_: Exception) { null }
    }

    @Synchronized
    fun downloadAndCache(desktop: JSONObject, descriptor: JSONObject): SchemeCacheResult {
        val desktopId = desktop.getString("desktopId")
        val version = descriptor.optString("version", "0")
        val hash = descriptor.optString("hash", "")
        val hasScheme = descriptor.optBoolean("hasScheme", false)
        val existing = get(desktopId)
        if (hash.isNotBlank() && existing?.optString("hash") == hash) {
            return SchemeCacheResult(false, hasScheme, version, hash)
        }

        val payload = if (hasScheme) downloadSchemePayload(desktop, descriptor) else JSONObject()
            .put("activeSchemeId", JSONObject.NULL)
            .put("scheme", JSONObject.NULL)
            .put("pages", JSONArray())
            .put("components", JSONArray())
            .put("actions", JSONArray())
            .put("permissionGrants", JSONArray())
        val assetRoot = schemeAssetRoot(desktopId, hash)
        try {
            materializeAssets(desktop, assetRoot, payload)
        } catch (error: Exception) {
            assetRoot.deleteRecursively()
            throw error
        }
        val cache = JSONObject()
            .put("desktopId", desktopId)
            .put("version", version)
            .put("hash", hash)
            .put("updatedAt", Instant.now().toString())
            .put("activeSchemeId", payload.opt("activeSchemeId"))
            .put("scheme", payload.opt("scheme"))
            .put("pages", payload.optJSONArray("pages") ?: JSONArray())
            .put("components", payload.optJSONArray("components") ?: JSONArray())
            .put("actions", payload.optJSONArray("actions") ?: JSONArray())
            .put("permissionGrants", payload.optJSONArray("permissionGrants") ?: JSONArray())
        check(prefs.edit().putString("scheme:$desktopId", cache.toString()).commit()) {
            "无法提交方案缓存索引"
        }
        removeOldAssetDirectories(desktopId, hash)
        return SchemeCacheResult(true, hasScheme, version, hash)
    }

    private fun downloadSchemePayload(desktop: JSONObject, descriptor: JSONObject): JSONObject {
        val totalBytes = descriptor.optLong("totalBytes", 0)
        val expectedHash = descriptor.getString("hash")
        if (totalBytes <= 0 || totalBytes > 32L * 1024 * 1024) {
            throw IllegalStateException("方案快照大小无效")
        }
        val output = ByteArrayOutputStream(totalBytes.toInt())
        var offset = 0L
        while (offset < totalBytes) {
            val response = gateway.request(
                desktop.getString("host"),
                desktop.optInt("port", 48320),
                gateway.authorizedRequest("scheme-chunk", desktop.getString("trustCredential"))
                    .put("hash", expectedHash)
                    .put("offset", offset)
                    .put("length", 24 * 1024),
                expectedFingerprint = desktop.optString("gatewayFingerprint"),
            )
            if (!response.optBoolean("ok")) throw IllegalStateException(response.optString("message", "方案分块下载失败"))
            val chunk = response.getJSONObject("payload")
            val bytes = android.util.Base64.decode(chunk.getString("data"), android.util.Base64.DEFAULT)
            if (chunk.optLong("offset", -1) != offset || bytes.isEmpty() && offset < totalBytes) {
                throw IllegalStateException("方案分块顺序无效")
            }
            output.write(bytes)
            offset += bytes.size
        }
        val content = output.toByteArray()
        if (sha256(content) != expectedHash) throw IllegalStateException("方案完整性校验失败")
        return JSONObject(String(content, Charsets.UTF_8))
    }

    private fun materializeAssets(desktop: JSONObject, assetRoot: File, payload: JSONObject) {
        assetRoot.mkdirs()
        val pages = payload.optJSONArray("pages") ?: JSONArray()
        for (index in 0 until pages.length()) {
            val page = pages.optJSONObject(index) ?: continue
            replaceMediaSource(desktop, assetRoot, "page", page.optString("id"), page, "backgroundMediaSource")
        }
        val components = payload.optJSONArray("components") ?: JSONArray()
        for (index in 0 until components.length()) {
            val bundle = components.optJSONObject(index) ?: continue
            val definition = bundle.optJSONObject("definition") ?: continue
            val config = bundle.optJSONObject("visualConfig") ?: continue
            val background = config.optJSONObject("background")
            if (background != null) replaceMediaSource(desktop, assetRoot, "component", definition.optString("id"), background, "mediaSource")
            val image = config.optJSONObject("image")
            if (image != null) replaceMediaSource(desktop, assetRoot, "component", definition.optString("id"), image, "source")
        }
    }

    private fun replaceMediaSource(
        desktop: JSONObject,
        assetRoot: File,
        ownerKind: String,
        ownerId: String,
        target: JSONObject,
        key: String,
    ) {
        // JSONObject.optString 会把 JSON null 转成字面量 "null"，必须先按真实类型过滤。
        val source = (target.opt(key) as? String)?.trim().orEmpty()
        if (source.isBlank() || ownerId.isBlank()) return
        val fileName = try { Uri.parse(source).lastPathSegment.orEmpty() } catch (_: Exception) { File(source).name }
        if (fileName.isBlank()) return
        try {
            val file = downloadAsset(desktop, assetRoot, ownerKind, ownerId, fileName)
            target.put(key, Uri.fromFile(file).toString())
        } catch (error: Exception) {
            logs.append("Error", "SchemeAsset", "$ownerKind/$ownerId/$fileName：${error.message ?: "资源下载失败"}")
            throw error
        }
    }

    private fun downloadAsset(desktop: JSONObject, root: File, ownerKind: String, ownerId: String, fileName: String): File {
        val safeName = "${ownerKind}-${ownerId}-${File(fileName).name}".replace(Regex("[^A-Za-z0-9._-]"), "_")
        val destination = File(root, safeName)
        if (destination.exists() && destination.length() > 0) return destination
        // 推送事件、连接刷新和前端主动刷新可能来自不同线程。外层缓存事务已串行化，
        // 这里仍使用唯一临时文件，避免异常恢复或未来多实例访问时互相覆盖下载内容。
        val temporary = File.createTempFile("$safeName.", ".tmp", root)
        FileOutputStream(temporary).use { output ->
            var offset = 0L
            var total = Long.MAX_VALUE
            while (offset < total) {
                val response = gateway.request(
                    desktop.getString("host"),
                    desktop.optInt("port", 48320),
                    gateway.authorizedRequest("asset", desktop.getString("trustCredential"))
                        .put("ownerKind", ownerKind)
                        .put("ownerId", ownerId)
                        .put("fileName", fileName)
                        .put("offset", offset)
                        .put("length", 24 * 1024),
                    timeoutMs = 12_000,
                    expectedFingerprint = desktop.optString("gatewayFingerprint"),
                )
                if (!response.optBoolean("ok")) throw IllegalStateException(response.optString("message", "资源下载失败"))
                val chunk = response.getJSONObject("payload")
                val bytes = android.util.Base64.decode(chunk.getString("data"), android.util.Base64.DEFAULT)
                total = chunk.getLong("totalBytes")
                if (chunk.getLong("offset") != offset || bytes.isEmpty() && offset < total) throw IllegalStateException("资源分块顺序无效")
                output.write(bytes)
                offset += bytes.size
                if (chunk.optBoolean("complete")) break
            }
        }
        try {
            SchemeAssetFileStore.commit(temporary, destination)
        } catch (error: Exception) {
            temporary.delete()
            throw IllegalStateException("无法提交资源缓存：${error.message ?: "文件替换失败"}", error)
        }
        return destination
    }

    private fun schemeAssetRoot(desktopId: String, schemeHash: String): File {
        val safeDesktopId = desktopId.replace(Regex("[^A-Za-z0-9._-]"), "_")
        val safeHash = schemeHash.ifBlank { "empty" }.replace(Regex("[^A-Fa-f0-9._-]"), "_")
        return File(context.filesDir, "scheme-assets/$safeDesktopId/$safeHash")
    }

    private fun removeOldAssetDirectories(desktopId: String, currentHash: String) {
        val safeDesktopId = desktopId.replace(Regex("[^A-Za-z0-9._-]"), "_")
        val safeHash = currentHash.ifBlank { "empty" }.replace(Regex("[^A-Fa-f0-9._-]"), "_")
        val desktopRoot = File(context.filesDir, "scheme-assets/$safeDesktopId")
        val directories = desktopRoot.listFiles()
            ?.filter(File::isDirectory)
            ?.map { SchemeAssetDirectory(desktopId, it.name) }
            .orEmpty()
        val stale = SchemeAssetRetentionPolicy.staleDirectories(desktopId, safeHash, directories)
            .map(SchemeAssetDirectory::schemeHash)
            .toSet()
        desktopRoot.listFiles()?.filter { it.isDirectory && it.name in stale }?.forEach { it.deleteRecursively() }
    }

    private fun sha256(bytes: ByteArray): String {
        return MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
    }
}
